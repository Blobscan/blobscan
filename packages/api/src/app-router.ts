import { z } from "@blobscan/zod";

import { publicProcedure } from "./procedures";
import { blobRouter } from "./routers/blob";
import { blobStoragesStateRouter } from "./routers/blob-storages-state";
import { blockRouter } from "./routers/block";
import { blockchainSyncStateRouter } from "./routers/blockchain-sync-state";
import { indexerRouter } from "./routers/indexer";
import { searchRouter } from "./routers/search";
import { statsRouter } from "./routers/stats";
import { transactionRouter } from "./routers/tx";
import { t } from "./trpc-client";

export const appRouter = t.router({
  block: blockRouter,
  tx: transactionRouter,
  blob: blobRouter,
  search: searchRouter,
  stats: statsRouter,
  indexer: indexerRouter,
  syncState: blockchainSyncStateRouter,
  blobStoragesState: blobStoragesStateRouter,
  healthcheck: publicProcedure
    .meta({
      openapi: {
        method: "GET",
        path: "/healthcheck",
        summary: "connection healthcheck.",
        tags: ["system"],
      },
    })
    .input(z.void())
    .output(z.string())
    .query(() => "yay!"),
});

// export type definition of API
export type AppRouter = typeof appRouter;
