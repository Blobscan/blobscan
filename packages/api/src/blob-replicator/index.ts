import { FlowProducer, Worker } from "bullmq";
import type { FlowChildJob, FlowJob, WorkerOptions } from "bullmq";
import path from "path";

import { createOrLoadBlobStorageManager } from "@blobscan/blob-storage-manager";
import { BlobStorage } from "@blobscan/db";

import { FLOW_PRODUCER_QUEUE, STORAGE_QUEUES } from "./config";
import type { BlobReplicationJobData } from "./types";

const WORKERS_DIR = "workers";

const STORAGE_WORKER_FILES: Record<BlobStorage, string> = {
  GOOGLE: path.join(__dirname, WORKERS_DIR, "gcs.js"),
  SWARM: path.join(__dirname, WORKERS_DIR, "swarm.js"),
  POSTGRES: path.join(__dirname, WORKERS_DIR, "postgres.js"),
};
const PROPAGATOR_WORKER_FILE = path.join(
  __dirname,
  WORKERS_DIR,
  "propagator-flow-producer.js"
);

const replicatorFlowProducer = new FlowProducer();

const BLOB_STORAGES = Object.values(BlobStorage);

export async function setUpBlobReplicationWorkers() {
  const blobStorageManager = await createOrLoadBlobStorageManager();
  const workerOptions: WorkerOptions = {
    useWorkerThreads: true,
  };

  const storageWorkers = BLOB_STORAGES.filter(
    blobStorageManager.hasStorage
  ).reduce<Record<BlobStorage, Worker>>((workers, storageName) => {
    const storageWorker = new Worker(
      STORAGE_QUEUES[storageName],
      STORAGE_WORKER_FILES[storageName],
      workerOptions
    );

    storageWorker.on("completed", (job) => {
      console.log(`Completed job ${job.id} in ${STORAGE_QUEUES[storageName]}`);
    });

    storageWorker.on("failed", (job, err) => {
      console.error(
        `Failed job ${job?.id} in ${STORAGE_QUEUES[storageName]}: ${err}`
      );
    });

    workers[storageName] = storageWorker;

    return workers;
  }, {} as Record<BlobStorage, Worker>);

  if (Object.values(storageWorkers).length) {
    const replicatorWorker = new Worker(
      FLOW_PRODUCER_QUEUE,
      PROPAGATOR_WORKER_FILE,
      workerOptions
    );
  }
}

async function createBlobReplicationJob(
  data: BlobReplicationJobData
): Promise<FlowJob> {
  const blobStorageManager = await createOrLoadBlobStorageManager();
  const versionedHash = data.versionedHash;

  const storageJobs: FlowChildJob[] = BLOB_STORAGES.filter(
    blobStorageManager.hasStorage
  ).map<FlowChildJob>((storage) => {
    const jobId = `${storage}-${versionedHash}`;

    return {
      name: jobId,
      queueName: STORAGE_QUEUES[storage],
      data,
      opts: {
        jobId,
      },
    };
  });

  const jobId = `${FLOW_PRODUCER_QUEUE}-${versionedHash}`;

  return {
    name: jobId,
    queueName: FLOW_PRODUCER_QUEUE,
    data,
    opts: {
      jobId,
    },
    children: storageJobs,
  };
}

export async function queueBlobForReplication(data: BlobReplicationJobData) {
  const blobFlowJob = await createBlobReplicationJob(data);

  replicatorFlowProducer.add(blobFlowJob);
}

export async function queueBlobsForReplication(
  blobReplicationData: BlobReplicationJobData[]
) {
  const blobReplicationFlowJobs = await Promise.all(
    blobReplicationData.map((data) => createBlobReplicationJob(data))
  );

  return replicatorFlowProducer.addBulk(blobReplicationFlowJobs);
}
