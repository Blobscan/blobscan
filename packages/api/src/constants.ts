export const DEFAULT_LIMIT = 50;
