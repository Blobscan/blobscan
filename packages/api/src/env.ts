import { env as blobPropagatorEnv } from "@blobscan/blob-propagator";
import { env as blobStorageManagerEnv } from "@blobscan/blob-storage-manager";
import { env as openTelemetryEnv } from "@blobscan/open-telemetry";
import {
  z,
  createEnv,
  presetEnvOptions,
  nodeEnvSchema,
  booleanSchema,
  maskSensitiveData,
} from "@blobscan/zod";

export const env = createEnv({
  envOptions: {
    server: {
      CHAIN_ID: z.coerce.number().positive().default(1),
      SECRET_KEY: z.string(),
      NODE_ENV: nodeEnvSchema.optional(),
      METRICS_ENABLED: booleanSchema.default("false"),
    },

    ...presetEnvOptions,
  },
  display(env) {
    console.log(
      `API configuration: secretKey: ${maskSensitiveData(env.SECRET_KEY)}`
    );
    blobPropagatorEnv.display();
    blobStorageManagerEnv.display();
    openTelemetryEnv.display();
  },
});
