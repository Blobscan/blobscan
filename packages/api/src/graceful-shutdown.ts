import { blobPropagator } from "@blobscan/blob-propagator";
import { prisma } from "@blobscan/db";

export async function gracefulShutdown() {
  const shutdownOps = [];

  if (blobPropagator) {
    shutdownOps.push(blobPropagator.close());
  }

  shutdownOps.push(prisma.$disconnect());

  return Promise.all(shutdownOps);
}
