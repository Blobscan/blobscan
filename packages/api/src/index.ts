import type { inferRouterInputs, inferRouterOutputs } from "@trpc/server";

import type { AppRouter } from "./app-router";

export { appRouter, type AppRouter } from "./app-router";
export { createTRPCContext, createTRPCInnerContext } from "./context";
export type {
  TRPCInnerContext as TRCInnerContext,
  TRPCContext,
} from "./context";

export * from "@trpc/server/adapters/express";

export { env } from "./env";
export { gracefulShutdown } from "./graceful-shutdown";
export { metricsHandler } from "./instrumentation";
export type {
  BlobStorage,
  Rollup,
  Metric,
  MetricHistogram,
  MetricHistogramBucket,
  Metrics,
  MetricsClient,
} from "@blobscan/db";

/**
 * Inference helpers for input types
 * @example type HelloInput = RouterInputs['example']['hello']
 **/
export type RouterInputs = inferRouterInputs<AppRouter>;

/**
 * Inference helpers for output types
 * @example type HelloOutput = RouterOutputs['example']['hello']
 **/
export type RouterOutputs = inferRouterOutputs<AppRouter>;
