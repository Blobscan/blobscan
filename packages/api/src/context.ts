import { TRPCError } from "@trpc/server";
import type { inferAsyncReturnType } from "@trpc/server";
import type { CreateNextContextOptions } from "@trpc/server/adapters/next";
import type {
  NodeHTTPCreateContextFnOptions,
  NodeHTTPRequest,
  NodeHTTPResponse,
} from "@trpc/server/adapters/node-http";
import jwt from "jsonwebtoken";

import { blobPropagator } from "@blobscan/blob-propagator";
import { getBlobStorageManager } from "@blobscan/blob-storage-manager";
import { prisma } from "@blobscan/db";

import { env } from "./env";

export type CreateContextOptions =
  | NodeHTTPCreateContextFnOptions<NodeHTTPRequest, NodeHTTPResponse>
  | CreateNextContextOptions;

type CreateInnerContextOptions = Partial<CreateContextOptions> & {
  apiClient: string | null;
};

function getJWTFromRequest(
  req: NodeHTTPRequest | CreateNextContextOptions["req"]
) {
  const authHeader = req.headers.authorization;
  if (!authHeader) {
    return null;
  }

  try {
    const [type, token] = authHeader.split(" ");
    if (type !== "Bearer" || !token) {
      return null;
    }

    const decoded = jwt.verify(token, env.SECRET_KEY) as string;

    return decoded;
  } catch (err) {
    if (err instanceof jwt.JsonWebTokenError) {
      return null;
    }

    throw new TRPCError({ code: "BAD_REQUEST" });
  }
}

export async function createTRPCInnerContext(opts?: CreateInnerContextOptions) {
  const blobStorageManager = await getBlobStorageManager();

  return {
    prisma,
    blobStorageManager,
    blobPropagator,
    apiClient: opts?.apiClient,
  };
}

export function createTRPCContext(
  {
    scope,
  }: {
    scope: string;
  } = { scope: "" }
) {
  return async (opts: CreateContextOptions) => {
    try {
      const apiClient = getJWTFromRequest(opts.req);

      const innerContext = await createTRPCInnerContext({
        apiClient,
      });

      return {
        ...innerContext,
        scope,
        req: opts.req,
        res: opts.res,
      };
    } catch (err) {
      const err_ = err as Error;

      throw new TRPCError({
        code: "INTERNAL_SERVER_ERROR",
        message: `Failed to create TRPC context: ${err_.message}`,
      });
    }
  };
}

export type TRPCContext = inferAsyncReturnType<
  ReturnType<typeof createTRPCContext>
>;
export type TRPCInnerContext = inferAsyncReturnType<
  typeof createTRPCInnerContext
>;
