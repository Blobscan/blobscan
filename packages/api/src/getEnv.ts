import { env } from "@blobscan/env";

import { publicProcedure } from "./procedures";

export const getEnv = publicProcedure.query(() => {
  const clientEnv = Object.entries(env)
    .filter(([key]) => key.startsWith("NEXT_"))
    .map(([key, value]) => [key.replace(/^NEXT_/, ""), value]);

  return Object.fromEntries(clientEnv);
});
