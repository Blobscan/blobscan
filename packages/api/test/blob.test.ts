import type { inferProcedureInput } from "@trpc/server";
import { beforeAll, describe, expect, it } from "vitest";

import type { AppRouter } from "../src/app-router";
import { appRouter } from "../src/app-router";
import type { TRPCContext } from "../src/context";
import {
  createTestContext,
  runExpandsTestsSuite,
  runFiltersTestsSuite,
  runPaginationTestsSuite,
} from "./helpers";

type GetByIdInput = inferProcedureInput<AppRouter["blob"]["getByBlobId"]>;

describe("Blob router", async () => {
  let caller: ReturnType<typeof appRouter.createCaller>;
  let ctx: TRPCContext;

  beforeAll(async () => {
    ctx = await createTestContext();
    caller = appRouter.createCaller(ctx);
  });

  describe("getAll", () => {
    runPaginationTestsSuite("blob", (paginationInput) =>
      caller.blob.getAll(paginationInput).then(({ blobs }) => blobs)
    );

    runFiltersTestsSuite("blob", (baseGetAllInput) =>
      caller.blob.getAll(baseGetAllInput).then(({ blobs }) => blobs)
    );

    runExpandsTestsSuite("blob", ["block", "transaction"], (input) =>
      caller.blob.getAll(input).then(({ blobs }) => blobs)
    );
  });

  describe("getByBlobId", () => {
    it("should get a blob by versioned hash", async () => {
      const input: GetByIdInput = {
        id: "blobHash004",
      };

      const result = await caller.blob.getByBlobId(input);

      expect(result).toMatchSnapshot();
    });

    it("should get a blob by kzg commitment", async () => {
      const input: GetByIdInput = {
        id: "commitment004",
      };

      const result = await caller.blob.getByBlobId(input);

      expect(result).toMatchSnapshot();
    });

    it("should fail when trying to get a blob by a non-existent hash", async () => {
      await expect(
        caller.blob.getByBlobId({
          id: "nonExistingHash",
        })
      ).rejects.toMatchSnapshot();
    });

    it("should fail when getting a blob and the blob data is not available", async () => {
      const blobHash = "blobHash006";

      await ctx.prisma.blobData.delete({
        where: {
          id: blobHash,
        },
      });

      await expect(
        caller.blob.getByBlobId({
          id: blobHash,
        })
      ).rejects.toMatchSnapshot();
    });
  });
});
