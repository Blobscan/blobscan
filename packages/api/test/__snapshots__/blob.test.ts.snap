// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Blob router > getAll > when getting paginated blob results > should default to the first page when no page was specified 1`] = `
[
  {
    "commitment": "0xb4f67eb0771fbbf1b06b88ce0e23383daf994320508d44dd30dbd507f598c0d9b3da5a152e41a0428375060c3803b983",
    "proof": "89cf91c4c8be6f2a390d4262425f79dffb74c174fb15a210182184543bf7394e5a7970a774ee8e0dabc315424c22df0f",
    "size": 1500,
    "versionedHash": "0x010001c79d78a76fb9b4bab3896ee3ea32f3e2607da7801eb1a92da39d6c1368",
  },
  {
    "commitment": "commitment001",
    "proof": "proof001",
    "size": 1000,
    "versionedHash": "blobHash001",
  },
]
`;

exports[`Blob router > getAll > when getting paginated blob results > should return the correct number of results 1`] = `
[
  {
    "commitment": "0xb4f67eb0771fbbf1b06b88ce0e23383daf994320508d44dd30dbd507f598c0d9b3da5a152e41a0428375060c3803b983",
    "proof": "89cf91c4c8be6f2a390d4262425f79dffb74c174fb15a210182184543bf7394e5a7970a774ee8e0dabc315424c22df0f",
    "size": 1500,
    "versionedHash": "0x010001c79d78a76fb9b4bab3896ee3ea32f3e2607da7801eb1a92da39d6c1368",
  },
  {
    "commitment": "commitment001",
    "proof": "proof001",
    "size": 1000,
    "versionedHash": "blobHash001",
  },
]
`;

exports[`Blob router > getByBlobId > should get a blob by kzg commitment 1`] = `
{
  "commitment": "commitment004",
  "data": "0xd76df869b71d79f835db7e39ebbf3d69b71d7e",
  "dataStorageReferences": [
    {
      "blobStorage": "POSTGRES",
      "dataReference": "blobHash004",
    },
    {
      "blobStorage": "GOOGLE",
      "dataReference": "7011893058/ob/Ha/sh/obHash004.txt",
    },
  ],
  "proof": "proof004",
  "size": 1300,
  "versionedHash": "blobHash004",
}
`;

exports[`Blob router > getByBlobId > should get a blob by versioned hash 1`] = `
{
  "commitment": "commitment004",
  "data": "0xd76df869b71d79f835db7e39ebbf3d69b71d7e",
  "dataStorageReferences": [
    {
      "blobStorage": "POSTGRES",
      "dataReference": "blobHash004",
    },
    {
      "blobStorage": "GOOGLE",
      "dataReference": "7011893058/ob/Ha/sh/obHash004.txt",
    },
  ],
  "proof": "proof004",
  "size": 1300,
  "versionedHash": "blobHash004",
}
`;
